File: next-right-pointer.php
<?php
/**
 * 116. Populating Next Right Pointers in Each Node PHP LeetCode
 * Definition for a Node.
 */ 
class Node {
    function __construct($val = 0) {
        $this->val = $val;
        $this->left = null;
        $this->right = null;
        $this->next = null;
    }
}

class Solution {
    public $root;
    function __construct()
    {
        $this->root = null;
    }

    function insert($data)
    {
        if( $this->root == null )
            $this->root = new Node($data);
        else {
            $current = $this->root;
            while(true) {
                if( $data < $current->val ) {
                    if( ! $current->left ){
                        $current->left = new Node($data);
                        break;
                    }
                    else {
                        $current = $current->left;
                    }
                }
                else if( $data > $current->val ){
                    if( ! $current->right ) {
                        $current = $current->right;
                        break;
                    }
                    else {
                        $current = $current->right;
                    }
                }
                else {
                    break;
                }
            }
        }
    }

    function createTree($root=array())
    {
        foreach($root as $r)
            $this->insert($r);
    }
    /**
    @param Node $root
    @return Node
     */
    public function connect($root) {
        if($root==null)
            return;

        $queue = [$root];
        while( ! empty($queue)){
            $prev = null;
            foreach( $queue as $q ){
                if( $prev && ! $prev->next && $q->left)
                    $prev->next = $q->left;
                else if( $prev && ! $prev->next )
                    $prev->next = '#';

                if( $q->left && $q->right )
                    $q->left->next = $q->right;
                if( $q->left && ! $q->right )
                    $q->left->next = '#';
                if($q->right)
                    $prev = $q->right;
                else if($q->left)
                    $prev = $q->left;
                else    
                    $prev = null;
            }
            $depth++;
            $next_level = [];
            foreach( $queue as $q ){
                if( $q->left ) {
                    $next_level[] = $q->left;
                }
                if( ! $q->left && $q->right )
                    $next_level[] = new Node(0);
                if( $q->left && ! $q->right )
                    $next_level[] = new Node(0);
                if( $q->right )
                    $next_level[] = $q->right;
            }
            $queue = $next_level;
        }
        return $root;
    }
}

$root = [1,2,3,4,5,6,7];

$sol = new Solution();
//$tree = $sol->createTree($root);
$r = new Node(1);
$r->left = new Node(2);
$r->left->left = new Node(4);
$r->left->right = new Node(5);
$r->right = new Node(3);
$r->right->left = new Node(6);
$r->right->right = new Node(7);
$ret = $sol->connect($r);
print_r( $ret );

File: next-right-pointer.py
"""
# 116. Populating Next Right Pointers in Each Node Python LeetCode

# Definition for a Node.
"""
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next


class Solution:
    def __init__(self):
        self.head = null
    
    def push(self, data:int )-> None:
        if not self.head:
            self.head = Node(data)
        else:
            current = self.head
            while(True):
                if current.val < data:
                    if not current.left:
                        current.left = Node(data)
                        break
                    else:
                        current = current.left
                elif current.val > data:
                    if not current.right:
                        current.right = Node(data)
                        break
                    else:
                        current = current.right
                else:
                    break
                
    def connect(self, root: Optional[Node]) -> Optional[Node]:
        if not root:
            return

        queue = [root]
        while( queue ):
            prev = None
            for q in queue:
                if prev and not prev.next and q.left:
                    prev.next = q.left
                elif prev and not prev.next and not q.left:
                    prev.next = '#'

                if q.left and q.right:
                    q.left.next = q.right
                if q.left and not q.right:
                    q.left.next = '#'

                if q.right:
                    prev = q.right
                elif q.left:
                    prev = q.left
                else:
                    prev = None

            next_level = []
            for q in queue:
                if q.left:
                    next_level.append(q.left)
                if q.left and not q.right:
                    next_level.append(Node(0))
                if not q.left and q.right:
                    next_level.append(Node(0))
                if q.right:
                    next_level.append(q.right)
            queue = next_level
        return root